bytecomp/bytegen.ml

```diff
@@ -747,6 +798,10 @@ let rec comp_expr stack_info env exp sz cont =
               comp_expr stack_info new_env exp sz
                 (Kpush :: Kacc i :: Kccall("caml_update_dummy", 2) ::
                  comp_rec new_env sz (i-1) rem)
+          | (_id, exp, RHS_unreachable)
+            :: rem ->
+              comp_expr stack_info new_env exp sz
+                (discard_dead_code (comp_rec new_env sz (i-1) rem))
```

Comment:
In the native backend you use exactly the same code generation for RHS_nonrec and RHS_unreachable toplevel definitions. Here in bytecode you handle then closer to the recursive definitions. Which approach is the right one?

I wonder if handling unreachable cases (really: aborting, non-returning) like nonrec cases could create issues of scoping in generated code for examples like `let rec foo = (ignore bar; assert false) and bar = ()`.

Reply:
I can't remember why I didn't just handle `RHS_unreachable` as `RHS_nonrec` when I wrote the bytecode patch. It does look safer that way, and I certainly don't want to introduce useless optimizations for these cases.

Reply:
I am not completely sure that it is okay to always treat `_unreachable` as a non-recursive binding: morally I could see someone defining a binding that first ignores a mutually-recursive name and then raises/aborts. But I cannot get any such example to work, because all constructions that only raise/abort (including matching on an un-inhabited GADT with only a refutation clause) seem to be marked Dynamic in the rec_check, and therefore reject any use of a recursive name.

Given this observation, it does seem reasonable to assume that all `_unreachable` toplevel values are non-recursive.

Reply:
Doesn't `let rec x = let _ = x in let 0 = 1 in 0` match your definition ? It always raises, uses its recursive binding, and is not rejected by `Rec_check`.
But anyway, `RHS_nonrec` already contains a lot of things that were considered `Static` by `Rec_check`, so it actually has to handle those cases more or less correctly.